presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/BaseCostBasedPlanTest.java

```diff
@@ -0,0 +1,248 @@
+        plan.getRoot().accept(joinOrderPrinter, 0);
+        return joinOrderPrinter.result();
+    }
+
+    private static class JoinOrderPrinter
+            extends SimplePlanVisitor<Integer>
+    {
+        private final StringBuilder result = new StringBuilder();
+
+        public String result()
+        {
+            return result.toString();
+        }
+
+        @Override
+        public Void visitJoin(JoinNode node, Integer indent)
+        {
+            JoinNode.DistributionType distributionType = node.getDistributionType()
+                    .orElseThrow(() -> new IllegalStateException("Expected distribution type to be set"));
+            if (node.isCrossJoin()) {
+                checkState(node.getType() == INNER && distributionType == REPLICATED, "Expected CROSS JOIN to be INNER REPLICATED");
+                output(indent, "cross join:");
+            }
+            else {
+                output(indent, "join (%s, %s):", node.getType(), distributionType);
+            }
+
+            return visitPlan(node, indent + 1);
+        }
+
+        @Override
+        public Void visitExchange(ExchangeNode node, Integer indent)
+        {
+            Partitioning partitioning = node.getPartitioningScheme().getPartitioning();
+            output(
+                    indent,
+                    "%s exchange (%s, %s, %s)",
+                    node.getScope().name().toLowerCase(ENGLISH),
+                    node.getType(),
+                    partitioning.getHandle(),
+                    partitioning.getArguments().stream()
+                            .map(Object::toString)
+                            .sorted() // Currently, order of hash columns is not deterministic
+                            .collect(joining(", ", "[", "]")));
+
+            return visitPlan(node, indent + 1);
+        }
+
+        @Override
+        public Void visitAggregation(AggregationNode node, Integer indent)
+        {
+            output(
+                    indent,
+                    "%s aggregation over(%s)",
+                    node.getStep().name().toLowerCase(ENGLISH),
+                    node.getGroupingKeys().stream()
+                            .map(Object::toString)
+                            .sorted()
+                            .collect(joining(", ")));
+
+            return visitPlan(node, indent + 1);
+        }
+
+        @Override
+        public Void visitTableScan(TableScanNode node, Integer indent)
+        {
+            output(indent, "%s", node.getTable().getConnectorHandle());
+            return null;
+        }
+
+        @Override
+        public Void visitSemiJoin(final SemiJoinNode node, Integer indent)
+        {
+            output(indent, "semijoin (%s):", node.getDistributionType().get());
+
+            return visitPlan(node, indent + 1);
+        }
+
+        @Override
+        public Void visitValues(ValuesNode node, Integer indent)
+        {
+            output(indent, "values");
```

Comment:
I'm wondering if it makes sense to print the values here? or at least the size. There could be multiple value nodes in the plan.